pkg/controller/machine-set-boot-image/platform_helpers.go

```diff
@@ -192,9 +212,485 @@ func reconcileLibvirt(machineSet *machinev1beta1.MachineSet, _ *corev1.ConfigMap
+	if err != nil {
+		return "", fmt.Errorf("unable to get tags for category: %w", err)
+	}
+	for i, tag := range categoryTags {
+		if tag.Name == clusterID {
+			categoryTag = &categoryTags[i]
+			tagID = tag.ID
+			break
+		}
+	}
+
+	if categoryTag == nil {
+		categoryTag = &tags.Tag{
+			Description: "Added by openshift-install do not remove",
+			Name:        clusterID,
+			CategoryID:  tagCategoryID,
+		}
+		tagID, err = tagManager.CreateTag(ctx, categoryTag)
+		if err != nil {
+			return "", fmt.Errorf("unable to create tag: %w", err)
+		}
+	}
+
+	return tagID, nil
+}
+
+// Used govc/importx/ovf.go as an example to implement
+// resourceVspherePrivateImportOvaCreate and upload functions
+// See: https://github.com/vmware/govmomi/blob/cc10a0758d5b4d4873388bcea417251d1ad03e42/govc/importx/ovf.go#L196-L324
+func upload(ctx context.Context, archive *importer.TapeArchive, lease *nfc.Lease, item nfc.FileItem) error {
+	file := item.Path
+
+	f, size, err := archive.Open(file)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	opts := soap.Upload{
+		ContentLength: size,
+	}
+
+	return lease.Upload(ctx, item, f, opts)
+}
+
+func createNewVMTemplateWithName(streamData *stream.Stream, providerSpec *machinev1beta1.VSphereMachineProviderSpec, infra *osconfigv1.Infrastructure, arch string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Find and download the relevant OVA file
+	ova, err := streamData.QueryDisk(arch, "vmware", "ova")
+	if err != nil {
+		return "", err
+	}
+
+	klog.Infof("Downloading %s\n", ova.Location)
+	ovaPath, err := ova.Download(".")
+	if err != nil {
+		return "", fmt.Errorf("Failed to download %s: %w", ova.Location, err)
+	}
+	klog.Infof("Downloaded %s\n", ovaPath)
+
+	var name string
+	for _, vcenter := range infra.Spec.PlatformSpec.VSphere.VCenters {
+		if vcenter.Server != providerSpec.Workspace.Server {
+			continue
+		}
+		vcenterURL, err := url.Parse(vcenter.Server)
+		if err != nil {
+			return "", err
+		}
+
+		client, err := govmomi.NewClient(ctx, vcenterURL, true)
+		if err != nil {
+			return "", err
+		}
+
+		restClient := rest.NewClient(client.Client)
+		err = restClient.Login(ctx, vcenterURL.User)
+		if err != nil {
+			logoutErr := client.Logout(context.TODO())
+			if logoutErr != nil {
+				err = logoutErr
+			}
+			return "", err
+		}
+
+		// Initialize tagging client
+		tagManager := tags.NewManager(restClient)
+
+		finder := find.NewFinder(client.Client, false)
+
+		tagID, err := createClusterTagID(ctx, tagManager, infra.Status.InfrastructureName)
+		if err != nil {
+			return "", err
+		}
+
+		for _, failureDomain := range infra.Spec.PlatformSpec.VSphere.FailureDomains {
+			if failureDomain.Server != vcenter.Server {
+				continue
+			}
+
+			name = fmt.Sprintf("%s-rhcos-%s-%s", infra.Status.InfrastructureName, failureDomain.Region, failureDomain.Zone)
+
+			folderPath := providerSpec.Workspace.Folder
```

Comment:
hmm, how are you determining which failure domain is the correct machine ?

Reply:
by ensuring (failureDomain.Server and vcenter.Server) == providerSpec.Workspace.Server